### DISCUSSION TITLE
Using MOSAIC and SUMO in docker containers - can't get them to establish communication

### DISCUSSION THREAD

**Discussion by dalybrown** [eclipse/mosaic#322] [Q&A] [answered]
Hello,

I'm trying to run the MOSAIC tutorials (specifically, the Barnim one) except I am running MOSAIC and SUMO in separate docker containers. MOSAIC seems to throw an exception as soon as it tries to connect to SUMO:

Excerpt from the `MOSAIC.log` file:

```
2023-05-23 17:54:36,372 ERROR MosaicStarter - Stopping simulation due to a critical error.
org.eclipse.mosaic.rti.api.InternalFederateException: Could not initialize SUMO ambassador. Please see Traffic.log for details.
...
Caused by: java.io.IOException: Could not load establish connection to SUMO due to an unknown error.
...
Caused by: org.eclipse.mosaic.rti.api.InternalFederateException: Error during executing TraCI command SimulationGetVersion
...
Caused by: java.net.SocketException: Broken pipe
```

Excerpt from the `Traffic.log` file:

```
...
2023-05-23 17:54:35,494 DEBUG SumoAmbassador - Created TraCI Connection with the following specs: 
2023-05-23 17:54:35,494 DEBUG SumoAmbassador -     Receive Buffer Size: 65536
2023-05-23 17:54:35,494 DEBUG SumoAmbassador -     Send Buffer Size: 43520
2023-05-23 17:54:35,494 DEBUG SumoAmbassador -     Keep alive: false
2023-05-23 17:54:35,494 DEBUG SumoAmbassador -     TCP NoDelay: true
...
2023-05-23 17:54:36,162 ERROR TraciClientBridge - Close all TraCI streams due to an error
java.net.SocketException: Broken pipe
...
2023-05-23 17:54:36,162 ERROR SumoAmbassador - Error while trying to initialize SUMO ambassador.
...
java.io.IOException: Could not load establish connection to SUMO due to an unknown error.
...
Caused by: org.eclipse.mosaic.rti.api.InternalFederateException: Error during executing TraCI command SimulationGetVersion
...
Caused by: java.net.SocketException: Broken pipe
2023-05-23 17:54:36,162 INFO  SumoAmbassador - Closing SUMO connection
2023-05-23 17:54:36,162 INFO  SumoAmbassador - Finished simulation
```

I'm using version `23.0` of MOSAIC and `1.16` of SUMO.

SUMO doesn't seem to output anything important in its log:

```
***Starting server on port 9999 ***
Quitting (on error).
```

Does anyone have any ideas? I'm updated the `hosts.json` and `runtime.json` to the best of my ability. I can ping the containers (from within them) using the hostname instead of the IP address... so they can reach each other (which seems to be the case because SUMO immediately quits when MOSIAC tries to connect). 

I am using the following Java version:

```
openjdk 17.0.7 2023-04-18
OpenJDK Runtime Environment Temurin-17.0.7+7 (build 17.0.7+7)
OpenJDK 64-Bit Server VM Temurin-17.0.7+7 (build 17.0.7+7, mixed mode)
```

And the following SUMO version:

```
Eclipse SUMO sumo Version 1.16.0
 Build features: Linux-5.15.49-linuxkit aarch64 GNU 10.2.1 Release FMI Proj GUI Intl SWIG GDAL GL2PS Eigen
 Copyright (C) 2001-2023 German Aerospace Center (DLR) and others; https://sumo.dlr.de
 License EPL-2.0: Eclipse Public License Version 2 <https://eclipse.org/legal/epl-v20.html>
 Use --help to get the list of options.
```

Thanks!

**Comment by schwepmo** [comment, marked as answer]
I think this issue is related to the fact that MOSAICs `SumoAmbassador` is trying to start SUMO locally. This is required as we need to write vehicle types from the `mapping_config.json` to a SUMO configuration file as it is not possible to register new types via TraCI. Unfortunately, I believe there is no "simple" way to achieve what you're trying to do. I suppose you could try to implement your own `RemoteSumoAmbassador` by extending the `AbstractSumoAmbassador` and adjusting the `initSumoConnection()`-method.

Sorry for this *bad* news, maybe @kschrab can extend further on my answer.

**Reply by dalybrown** [reply]
Thanks for the reply.

I noticed that there were `start` and `deploy` parameters in the `runtime.json` file for SUMO. I set those to false because I would have started SUMO already. I guess there is where I assumed MOSAIC was not trying to start SUMO.

I guess I don't know enough about the two applications yet, I assumed the regular `SumoAmbassador` was the *remote* one since it uses a socket connection. From what I understand of your answer, even if it uses a socket connection, it still needs to write some configuration file that is visible to SUMO (or both applications).

I'll await the further extension of your answer but I'll try something else to see if I can get it working. 

As a side note, is there additional logging for SUMO I can turn up? I haven't figured out how yet so it's tough to debug if my issue is on the MOSAIC side or the SUMO side.

Thanks!

**Reply by schwepmo** [reply]
Yes exactly, unfortunately in order for us to define custom vehicle types we need to write the additional *mosaic_types.add.xml* file ([further described here](https://www.eclipse.org/mosaic/docs/simulators/traffic_simulator_sumo/#deep-dive-route-files--additional-files))

By "additional logging", do you mean additional to what is output in the *Traffic.log* file?

**Reply by dalybrown** [reply]
Sorry ... by "additional logging" I meant in SUMO ... I haven't figured out how to turn on any sort of verbose logging in that application to help debug.

There seems to be plenty of logging in MOSIAC; however, it wasn't clear to me that the the reason the socket connection couldn't be established was for the reason you mentioned.

Maybe in the future, when I have some free time, I'll try to implement what you suggested or come up with a way to get the deployment architecture I want working and contribute it back. For now, I'll follow your suggestions and run them in the same container.

Thanks for the help!
